packages/eslint-plugin/src/rules/type-annotation-spacing.ts

```diff
@@ -1,22 +1,37 @@
-import { TSESTree } from '@typescript-eslint/experimental-utils';
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
 import * as util from '../util';
 
-type Options = [
-  {
-    before?: boolean;
-    after?: boolean;
-    overrides?: {
-      colon?: {
-        before?: boolean;
-        after?: boolean;
-      };
-      arrow?: {
-        before?: boolean;
-        after?: boolean;
-      };
-    };
-  }?,
-];
+interface WhitespaceRule {
+  readonly before?: boolean;
+  readonly after?: boolean;
+}
+
+interface WhitespaceOverride {
+  readonly colon?: WhitespaceRule;
+  readonly arrow?: WhitespaceRule;
+  readonly variable?: WhitespaceRule;
+  readonly property?: WhitespaceRule;
+  readonly parameter?: WhitespaceRule;
+  readonly returnType?: WhitespaceRule;
+}
+
+interface Config extends WhitespaceRule {
+  readonly overrides?: WhitespaceOverride;
+}
+
+interface WhitespaceRules {
+  colon: WhitespaceRule;
+  arrow: WhitespaceRule;
+  variable: WhitespaceRule;
+  parameter: WhitespaceRule;
+  property: WhitespaceRule;
+  returnType: WhitespaceRule;
+}
```

Comment:
nit, could represent this as a mapped type to reduce duplication
```suggestion
type WhitespaceRules = { [k in keyof WhitespaceOverride]-?: WhitespaceRule }
```

Reply:
Instead I made WhitespaceRules a Required<WhitespaceOverride>. Hope that works too.